book/src/dev/rfcs/0006-stolon.md

```diff
@@ -0,0 +1,399 @@
+
+To handle fallback, a Zebra node without Tor access can perform the relay
+phase over an existing peer connection, rather than over an ephemeral Tor
+connection. This is much worse for privacy, but is similar to the design of
+Dandelion-lite. Ideally, Zebra+Tor is the common case for Zcash nodes, and
+Zebra nodes with Tor support can improve network privacy for Zebra without
+Tor and Zcashd nodes.
+
+We should consider how we can encourage Tor deployments for Zebra. One option
+would be to recommend Docker as the primary install mechanism, and bundle a Tor
+relay image with the Zebra image. Another would be to attempt to package Zebra
+with a dependency on Tor.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This design is more complicated than simply immediately broadcasting
+  transactions to all peers.
+
+- This design requires integration with external software (Tor) to be
+  effective, although the fallback method probably provides better privacy than
+  immediate broadcast.
+
+- Zcashd keeps IP-based peer reputation scores and may reject connections from
+  Tor exits if too many invalid transactions are sent from one exit.  However,
+  only one relay-to-Zcashd has to succeed for Zcashd to broadcast a transaction
+  to all of its peers, and Zebra, which does not yet have a reputation system,
+  can implement one that's aware of transaction relaying.  It's unclear whether
+  this will be a problem in practice, and the solution would probably be to fix
+  Zcashd.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Dandelion without Tor: this option has the advantage of not requiring
+  integration with external software, but it provides weaker privacy benefits.
+  Unlike Dandelion, in Stolon, a single step in the relay phase completely
+  unlinks a transaction from its source.
+
+* Overlap with wallet code: the code required to implement Stolon is reusable
+  for wallet software, and provides strong anonymity properties for that use
+  case regardless of how widely Zebra is deployed.
+
+* It's something: Zebra needs a transaction diffusion mechanism, and this one
+  can share code with wallet software and provide strong benefits in the
+  Zebra+Tor case, even if the privacy properties of the fallback mechanism are
+  unknown.
+
+* Using Tor connections only for wallets, not for node relay: this doesn't save
+  much complexity, since the Tor code has to exist anyways, but it does prevent
+  Zebra+Tor nodes from anonymously relaying transactions on behalf of software
+  that does not use Tor connections.
+
+* Guard peers: instead of using random clearnet peers for non-Tor fallback, we
+  could choose `f` guard peers as in Tor.  This is probably workable but would
+  complicate the backpressure design of the peer set somewhat, as we would
+  probably have to dedicate `f` peers to *only* serve transaction relay
+  requests.  This might introduce additional network distinguishers.
```

Comment:
Should we bake this into the design instead of using random relay peers? Using random relay peers is totally fine in the Zebra+Tor case, but is probably worse in the fallback case.  So that's a reason to try to implement guard peers.

On the flip side, though, using guard peers is more complicated from the implementation point of view.  And because we haven't done any kind of empirical or theoretical analysis on the privacy protections provided in the fallback case, although I expect that it would be better, I don't think we actually know that it will be.  So this might be a reason to do nothing for now, and move analysis of the fallback case + guard peers to future work. (See the "it's something" above).

Reply:
I think there are a number of risks with guard peers, which need further analysis:
* guard peer loops (not possible in the Tor case, because clients build the paths, and relays reject 2-relay loops)
* the guard set links all transactions that are broadcast by a particular peer
  * worse security for the Tor transport (?)
  * about the same for the non-Tor transport, because the source IP links them anyway

We should also consider a design where clients (wallets) choose the guards, rather than relays (nodes).

Ultimately, guards are a privacy tradeoff for a large set of clients with a small proportion of bad peers:
* random distribution ensures eventual compromise for all long-term clients
* guards protect most clients, but ensure compromise for a small proportion of clients

It's unclear if this tradeoff is the right one for typical Zcash threat models.

We also don't have enough analysis to choose the number of guards (2 ?), the guard rotation period (weeks? months?), and the guard replacement algorithm, for when a guard is not available.

Reply:
I think that the simplest thing to do would be not to use guard peers.  Since the baseline is Zcashd's immediate broadcast, we're not worse than the current state of affairs, and it lets us focus effort on making the Tor case the common case.